tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc

```diff
@@ -92,113 +45,115 @@ using namespace tensorpipe::channel;
+    TP_CUDA_CHECK(cudaMemsetAsync(ptr, 0x42, kSize, sendStream));
+
+    // Perform send and wait for completion.
+    auto descriptorPromise = std::make_shared<
+        std::promise<std::tuple<tensorpipe::Error, std::string>>>();
+    auto sendPromise = std::make_shared<std::promise<tensorpipe::Error>>();
+    auto descriptorFuture = descriptorPromise->get_future();
+    auto sendFuture = sendPromise->get_future();
+
+    channel->send(
+        CudaTensor{
+            .ptr = ptr,
+            .length = kSize,
+            .stream = sendStream,
+        },
+        [descriptorPromise{std::move(descriptorPromise)}](
+            const tensorpipe::Error& error, std::string descriptor) {
+          descriptorPromise->set_value(
+              std::make_tuple(error, std::move(descriptor)));
+        },
+        [sendPromise{std::move(sendPromise)}](const tensorpipe::Error& error) {
+          sendPromise->set_value(error);
+        });
+
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) = descriptorFuture.get();
+
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kClient, descriptor);
+    Error sendError = sendFuture.get();
+    EXPECT_FALSE(sendError) << sendError.what();
+    TP_CUDA_CHECK(cudaFree(ptr));
+
+    this->peers_->done(PeerGroup::kServer);
+    this->peers_->join(PeerGroup::kServer);
+
+    ctx->join();
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<CudaContext> ctx = this->helper_->makeContext("client");
+    auto channel = std::static_pointer_cast<cuda_ipc::Channel>(
+        ctx->createChannel(std::move(conn), Endpoint::kConnect));
+
+    TP_CUDA_CHECK(cudaSetDevice(0));
+    cudaStream_t recvStream;
+    TP_CUDA_CHECK(cudaStreamCreate(&recvStream));
+    void* ptr;
+    TP_CUDA_CHECK(cudaMalloc(&ptr, kSize));
+
+    auto descriptor = this->peers_->recv(PeerGroup::kClient);
+
+    // Perform recv and wait for completion.
+    auto recvPromise = std::make_shared<std::promise<tensorpipe::Error>>();
+    auto recvFuture = recvPromise->get_future();
+
+    channel->recv(
+        std::move(descriptor),
+        CudaTensor{
+            .ptr = ptr,
+            .length = kSize,
+            .stream = recvStream,
+        },
+        [recvPromise{std::move(recvPromise)}](const tensorpipe::Error& error) {
+          recvPromise->set_value(error);
+        });
+
+    Error recvError = recvFuture.get();
+    EXPECT_FALSE(recvError) << recvError.what();
+
+    std::array<uint8_t, kSize> data;
+    TP_CUDA_CHECK(cudaMemcpy(data.data(), ptr, kSize, cudaMemcpyDefault));
+    // Validate contents of vector.
+    for (auto i = 0; i < kSize; i++) {
+      EXPECT_EQ(data[i], 0x42);
+    }
+    TP_CUDA_CHECK(cudaFree(ptr));
+
+    this->peers_->done(PeerGroup::kClient);
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
+};
+
+CHANNEL_TEST(CudaIpc, ReceiverWaitsForStartEvent);
```

Comment:
I thought the first argument of `CHANNEL_TEST` is the type of the tensor, which should be `Cuda` here. In fact, this test (`ReceiverWaitsForStartEvent`) should probably go in a CUDA test suite, and then this CUDA test suite should be instantiated once for CudaIpc. Right?

Reply:
The first argument of `CHANNEL_TEST` is a prefix for a test suite, for instance `Foo` -> `FooChannelTest`. It coincides with the type of the tensor in the case of generic test suites, but that's just a convenience.

Reply:
Right, but (also relating to the comment above), this test isn't specific to the `cuda_ipc` channel, it could be run on all Cuda channels (and should work for all of them). So shouldn't we add it to the "generic" Cuda test suite, rather than to the one specific to this one channel?